### PR TITLE
[HTTP/3] Stress hack for msquic dropping connections

### LINKED ISSUE #77126: [HttpStress] massive timeouts around HTTP/3 stress test startup
This happened 3 times in the past 4 weeks, `main` and `release/7.0` are both affected:
- [09/24](https://dev.azure.com/dnceng-public/public/_build/results?buildId=28755&view=logs&j=0da5d1d9-276d-5173-c4c4-9d4d4ed14fdb&t=1451f5f3-0108-5a08-5b92-e984b2a85bbd)
- [09/26](https://dev.azure.com/dnceng-public/public/_build/results?buildId=30067&view=logs&j=0da5d1d9-276d-5173-c4c4-9d4d4ed14fdb&t=1451f5f3-0108-5a08-5b92-e984b2a85bbd)
- [10/10](https://dev.azure.com/dnceng-public/public/_build/results?buildId=46233&view=logs&j=22fea640-1099-5f32-ec5d-316ad83f359a&t=a923b061-8445-5c16-cd6e-848990587cb0)

Looks like sometimes the stress server is unresponsive for a long time around startup in these cases.

### PR DESCRIPTION
**Pull request by ManickaP**
Implement the same hack as for functional tests to prevent msquic from dropping connections, see #83687.

Fixes #77126

### CODE REVIEW

File: src/libraries/System.Net.Http/tests/StressTests/HttpStress/HttpStress.csproj [hunk 1]
```diff
@@ -4,8 +4,19 @@
...
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
   </PropertyGroup>
 
+  <!-- Shared production code -->
+  <ItemGroup>
+    <Compile Include="..\..\..\..\Common\src\System\Net\Logging\NetEventSource.Common.cs" Link="Common\System\Net\Logging\NetEventSource.Common.cs" />
+    <Compile Include="..\..\..\..\System.Net.Quic\src\System\Net\Quic\Internal\MsQuicApi.cs" Link="ProductionCode\System\Net\Quic\Internal\MsQuicApi.cs" />
+    <Compile Include="..\..\..\..\System.Net.Quic\src\System\Net\Quic\Internal\MsQuicSafeHandle.cs" Link="ProductionCode\System\Net\Quic\Internal\MsQuicSafeHandle.cs" />
+    <Compile Include="..\..\..\..\System.Net.Quic\src\System\Net\Quic\Interop\*.cs" Link="ProductionCode\System\Net\Quic\Interop\*.cs" />
+    <Compile Include="..\..\..\..\System.Net.Quic\src\System\Net\Quic\NetEventSource.Quic.cs" Link="ProductionCode\System\Net\Quic\NetEventSource.Quic.cs" />
```
Comment (antonfirsov) [inline src/libraries/System.Net.Http/tests/StressTests/HttpStress/HttpStress.csproj:17]: A refactor of these internals may break stress builds without us noticing, but I guess there is no good way to deal with this. We should remember that `HttpStress` took a dependency.

Comment (antonfirsov) [reply src/libraries/System.Net.Http/tests/StressTests/HttpStress/HttpStress.csproj:17]: Also: these files are missing from the docker context on Windows, which is why the Windows build failed. Copying them would make this solution even more fragile. Can we implement worker delay tuning hack with reflection instead and share that implementation between `System.Net.Quic.Functional.Tests` and the stress tests? That would also remove the need of including anything but `QuicDefaults.cs` in `System.Net.Quic.Functional.Tests.csproj`.

Comment (antonfirsov) [reply src/libraries/System.Net.Http/tests/StressTests/HttpStress/HttpStress.csproj:17]: Ah no, we would need #75347 for to invoke function pointers with reflection safely. I guess, the best would be to [copy](https://github.com/dotnet/runtime/blob/d27142772e8de8c47c97a7661f5bf03f1a2ff703/eng/docker/build-docker-sdk.ps1#L52-L53) & include `msquic_generated.cs`, then do a combination of reflection + code that only depends on the generated types.

I can take over if you want.

Comment (ManickaP) [reply src/libraries/System.Net.Http/tests/StressTests/HttpStress/HttpStress.csproj:17]: Reflection was one of the options we discussed with @CarnaViire when implementing this in functional tests. I was able to workaround it in functional tests with source code inclusion, which is normal for corresponding tests. However, here it proves problematic.
I could re-work this to create an internal function on `MsQuicApi` which we would invoke via reflection from both functional tests as well as stress.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicApi.cs [hunk 2]
```diff
@@ -227,4 +228,19 @@ private static bool IsTls13Disabled(bool isServer)
 #endif
         return false;
     }
+
+    // Do not change the name and signature without looking for textual occurrences!
+    // This method is invoked via reflection from QUIC functional and HTTP stress tests.
+    private static (bool, string) SetUpForTests()
```
[the remark this answers is not in the document]
Comment (ManickaP) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicApi.cs:234]: I haven't thought about that, but I found this for example:
https://github.com/dotnet/runtime/blob/2d833f43d4158c39ff62e57bf0f8d58b93935ae4/src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/HPackDecoder.cs#L109-L110

I can always go back to code sharing though and try to minimize the number of shared files, either with some some reflection, strategically placed `#if` and/or code stubs.

File: src/libraries/System.Net.Quic/tests/FunctionalTests/QuicListenerTests.cs [hunk 3]
```diff
@@ -119,7 +119,7 @@ public async Task AcceptConnectionAsync_SlowOptionsCallback_TimesOut(bool useCan
 
             // Connect attempt should be stopped with "UserCanceled".
             var connectException = await Assert.ThrowsAsync<AuthenticationException>(async () => await connectTask);
-            Assert.Contains(TlsAlertMessage.UserCanceled.ToString(), connectException.Message);
+            Assert.Contains("UserCanceled", connectException.Message);
```
Comment (rzikm) [inline src/libraries/System.Net.Quic/tests/FunctionalTests/QuicListenerTests.cs:122]: What is the motivation for this change? What about `nameof`?

Comment (ManickaP) [reply src/libraries/System.Net.Quic/tests/FunctionalTests/QuicListenerTests.cs:122]: It's an internal `enum` in security, it's not public. I removed all the shared source files, I haven't put any effort in figuring out why it was there. 
I see that the file with the enum doesn't have any dependencies, so it could be put back without dragging bunch of other stuff. I can do that.

Comment (ManickaP) [reply src/libraries/System.Net.Quic/tests/FunctionalTests/QuicListenerTests.cs]: In the end, we're bringing back some code sharing so I reverted this change.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicApi.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ManickaP) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicApi.cs:234]: Combo of shared code + reflection is there. @antonfirsov could you look into copying the appropriate files in the docker container please?

File: eng/docker/libraries-sdk.linux.Dockerfile [hunk 4]
```diff
@@ -1,5 +1,5 @@
 # Builds and copies library artifacts into target dotnet sdk image
-ARG BUILD_BASE_IMAGE=mcr.microsoft.com/dotnet-buildtools/prereqs:centos-7-f39df28-20191023143754
+ARG BUILD_BASE_IMAGE=mcr.microsoft.com/dotnet-buildtools/prereqs:centos-7
```
Comment (wfurt) [inline eng/docker/libraries-sdk.linux.Dockerfile:2]: 8.0 will not support Centos 7 AFAIK (hurray!) we should update this at some point

File: src/libraries/System.Net.Quic/tests/FunctionalTests/QuicTestBase.cs [hunk 5]
```diff
@@ -44,14 +44,20 @@ public abstract class QuicTestBase : IDisposable
 
         static unsafe QuicTestBase()
         {
-            Console.WriteLine($"MsQuic {(IsSupported ? "supported" : "not supported")} and using '{MsQuicApi.MsQuicLibraryVersion}'.");
+            // If any of the reflection bellow breaks due to changes in "System.Net.Quic.MsQuicApi", also check and fix HttpStress project as it uses the same hack.
```
Comment (wfurt) [inline src/libraries/System.Net.Quic/tests/FunctionalTests/QuicTestBase.cs:47]: I'm wondering if this should go somewhere to `Common/test/....` so we can avoid duplication.
It may be worth of adding explanation as this is unusual craft.

Comment (ManickaP) [reply src/libraries/System.Net.Quic/tests/FunctionalTests/QuicTestBase.cs:47]: Sounds good, I'll look into it.

File: eng/docker/libraries-sdk.linux.Dockerfile (hunk 4 above)
Comment (ManickaP) [reply eng/docker/libraries-sdk.linux.Dockerfile:2]: Stream9??? Or should we swap for different distro?

File: src/libraries/System.Net.Quic/tests/FunctionalTests/QuicTestBase.cs (hunk 5 above)
Comment (antonfirsov) [reply src/libraries/System.Net.Quic/tests/FunctionalTests/QuicTestBase.cs:47]: We should keep in mind that whatever utility helper we add, its! file has to be also copied to the container, so avoiding duplication has a higher cost in complexity here than normally. At the moment I'm undecided if it's worth it, would prefer to do the docker copying work mentioned in https://github.com/dotnet/runtime/pull/84793#discussion_r1172373409 first, planning to jump to it tomorrow.

Comment (ManickaP) [reply src/libraries/System.Net.Quic/tests/FunctionalTests/QuicTestBase.cs:47]: Ok, I'm holding back. Let me know when/if I should look into it.

File: eng/docker/libraries-sdk.linux.Dockerfile (hunk 4 above)
Comment (wfurt) [reply eng/docker/libraries-sdk.linux.Dockerfile:2]: the `centos-7` images have build tools and we use them for product. I don't know if that is needed. 
I think we should pick whatever is easiest for us to maintain.

Comment (antonfirsov) [reply eng/docker/libraries-sdk.linux.Dockerfile:2]: I'm switching to `centos-stream8` in #85342, which seems to work.

File: src/libraries/System.Net.Http/tests/StressTests/HttpStress/Program.cs [hunk 6]
```diff
@@ -158,6 +162,9 @@ private static async Task<ExitCode> Run(Configuration config)
 
             string GetAssemblyInfo(Assembly assembly) => $"{assembly.Location}, modified {new FileInfo(assembly.Location).LastWriteTime}";
 
+            Type msQuicApiType = typeof(QuicConnection).Assembly.GetType("System.Net.Quic.MsQuicApi");
+            string msQuicLibraryVersion = (string)msQuicApiType.GetProperty("MsQuicLibraryVersion", BindingFlags.NonPublic | BindingFlags.Static).GetGetMethod(true).Invoke(null, Array.Empty<object?>());
```
Comment (antonfirsov) [inline src/libraries/System.Net.Http/tests/StressTests/HttpStress/Program.cs:166]: @ManickaP unfortunately, this line fails with `NullReferenceException` in release builds. I assume the `MsQuicLibraryVersion` is being trimmed away, since it's unused in the production code.

We could workaround this by moving the version extraction to a shared test-utility (together with the MaxWorkerQueue hack helper), but IMHO it's not worth it unless we think that it's super-important to know the msquic version for stress test troubleshooting.

For now I'm just removing it.

Comment (wfurt) [reply src/libraries/System.Net.Http/tests/StressTests/HttpStress/Program.cs:166]: is it possible that there is image without MsQuic?

Comment (antonfirsov) [reply src/libraries/System.Net.Http/tests/StressTests/HttpStress/Program.cs:166]: What do you mean? If you are suspecting lack of msquic as the root cause of this error - it's not the case. Removing code messing with MsQuicLibraryVersion makes things work (= HTTP/3 stress runs).

Comment (ManickaP) [reply src/libraries/System.Net.Http/tests/StressTests/HttpStress/Program.cs:166]: Another option would be to tell the trimmer to keep it there via `DynamicDependency`. And this is just one string.

Comment (antonfirsov) [reply src/libraries/System.Net.Http/tests/StressTests/HttpStress/Program.cs:166]: We should try to clarify whether test-only usage of `DynamicDependency` is ok, without that, I prefer to defensively assume it's not.

File: src/libraries/System.Net.Http/tests/StressTests/HttpStress/build-local.ps1 [hunk 7]
```diff
@@ -14,7 +14,7 @@ if (-not ([string]::IsNullOrEmpty($args[0]))) {
 
 $LibrariesConfiguration = "Release"
 if (-not ([string]::IsNullOrEmpty($args[1]))) {
-    $LibrariesConfiguration = $args[0]
+    $LibrariesConfiguration = $args[1]
```
Comment (antonfirsov) [inline src/libraries/System.Net.Http/tests/StressTests/HttpStress/build-local.ps1:17]: Fixes in this file are unrelated. Both `$StressConfiguration` and `$LibrariesConfiguration` referred to `$args[0]` 🤦 .

File: src/libraries/System.Net.Http/tests/StressTests/HttpStress/Program.cs [hunk 8]
```diff
@@ -177,6 +182,22 @@ private static async Task<ExitCode> Run(Configuration config)
             Console.WriteLine("Query Parameters: " + config.MaxParameters);
             Console.WriteLine();
 
+            if (config.HttpVersion == HttpVersion.Version30 && IsQuicSupported)
+            {
+                Type msQuicApiType = typeof(QuicConnection).Assembly.GetType("System.Net.Quic.MsQuicApi");
```
Comment (antonfirsov) [inline src/libraries/System.Net.Http/tests/StressTests/HttpStress/Program.cs:187]: I would prefer to place a comment here about what is this code is doing and why, so it's obvious for someone new to the stress code.

File: src/libraries/System.Net.Quic/tests/FunctionalTests/System.Net.Quic.Functional.Tests.csproj [hunk 9]
```diff
@@ -31,18 +32,8 @@
...
-    <Compile Include="..\..\src\System\Net\Quic\Interop\*.cs" Link="ProductionCode\System\Net\Quic\Interop\*.cs" />
-    <Compile Include="..\..\src\System\Net\Quic\NetEventSource.Quic.cs" Link="ProductionCode\System\Net\Quic\NetEventSource.Quic.cs" />
     <Compile Include="..\..\src\System\Net\Quic\QuicDefaults.cs" Link="ProductionCode\System\Net\Quic\QuicDefaults.cs" />
-    <Compile Include="$(CommonPath)Interop\Windows\Interop.Libraries.cs" Link="Common\Interop\Windows\Interop.Libraries.cs" Condition="'$(TargetPlatformIdentifier)' == 'windows'" />
-    <Compile Include="$(CommonPath)Interop\Linux\Interop.Libraries.cs" Link="Common\Interop\Linux\Interop.Libraries.cs" Condition="'$(TargetPlatformIdentifier)' == 'linux'" />
-    <Compile Include="$(CommonPath)Interop\OSX\Interop.Libraries.cs" Link="Common\Interop\OSX\Interop.Libraries.cs" Condition="'$(TargetPlatformIdentifier)' == 'osx'" />
-    <Compile Include="$(CommonPath)System\Net\Security\TlsAlertMessage.cs" Link="Common\System\Net\Security\TlsAlertMessage.cs" />
+    <Compile Include="..\..\src\System\Net\Quic\Interop\*.cs" LinkBase="ProductionCode\System\Net\Quic\Interop" />
```
Comment (antonfirsov) [inline src/libraries/System.Net.Quic/tests/FunctionalTests/System.Net.Quic.Functional.Tests.csproj:36]: I changed this to use `LinkBase`, Visual Studio is going crazy when `Link` is being used with wildcards:

image

**Review by antonfirsov — COMMENTED** [review]
@ManickaP if you can address https://github.com/dotnet/runtime/pull/84793#discussion_r1178397960, I'm happy with the PR, if you are also happy with my changes. (Assuming the CI run succeeds.)

Note that copying the msquic interop stuff required to apply changes at 3 places:
- `build-docker-sdk.ps1`
- `Directory.Build.props`
- `windows.Dockerfile`

Given that, I think sharing more files between projects brings more maintenance burden than keeping the code duplication.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicApi.cs [hunk 10]
```diff
@@ -143,7 +143,7 @@ static MsQuicApi()
 
             if (NetEventSource.Log.IsEnabled())
             {
-                NetEventSource.Info(null, $"Loaded MsQuic library version '{version}', commit '{gitHash}'.");
+                NetEventSource.Info(null, $"Loaded MsQuic library '{MsQuicLibraryVersion}'.");
```
Comment (antonfirsov) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicApi.cs:146]: Brilliant 😆

**Comment by ManickaP** [comment]
If this passes, I think it's done. Thanks for the help @antonfirsov !

**Comment by ManickaP** [comment]
I'm seeing this build error in Linux runs:
```
/app/Directory.Build.targets(2,3): error MSB4019: The imported project "/live-runtime-artifacts/targetingpacks.targets" was not found. Confirm that the expression in the Import declaration "/live-runtime-artifacts/targetingpacks.targets" is correct, and that the file exists on disk. [/app/HttpStress.csproj]
```
Anything familiar Anton?

**Review by antonfirsov — APPROVED** [review]
This is good now. Fixed the linux build with [949d82b](https://github.com/dotnet/runtime/pull/84793/commits/949d82bfde443eb44f9ca753ef88a696eb741fe5). The failure in the last run is #76183.
